Move reflog into sqlite

ChrisPenner opened the pull request:
## Overview

Defines a format for keeping the reflog in sqlite;

* This ensures the reflog moves around with the codebase
* This guarantees that all references in the reflog refer to existing causals

(Ignore the weird duplications and such)

<img width="916" alt="image" src="https://user-images.githubusercontent.com/6439644/186757590-c9bfa549-ab76-4b6d-8bb7-9fcb22cfc373.png">

## Implementation notes

* Adds a new `reflog` table and an index on that table to sort by Date
* Adds a new migration which adds this table and ports an existing reflog to the db as best we can
* Rewrites the reflog output to use a table
* Note that the new reflog table only has (timestamp, new root causal hash, reason); we no longer track a `from` causal, since it will always just be the causal before it, and we don't show this in the output anyways.

When porting the old `reflog`:

* We throw out any malformed entries, or any entries which refer to causal hashes we don't have in the codebase
* We don't have timestamp info for old entries, so we just import them all at the "current time", but add second offsets such that they appear in the correct chronological order.
* I don't delete the old reflog file, I could, but there doesn't seem to be a good reason to and deleting files on the user's system is always a little bit sketch.

## Test coverage

Updated the reflog transcript

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/unisonweb/unison/3364)
<!-- Reviewable:end -->

ChrisPenner committed:
Add reflog table

ChrisPenner committed:
Add reflog table to sqlite

ChrisPenner committed:
Bubble up new reflog into codebase interface

ChrisPenner committed:
Add columnNHeader

ChrisPenner committed:
Switch from streaming to a simple query

ChrisPenner committed:
Add reflog db migration

ChrisPenner committed:
Fix reflog migration

ChrisPenner committed:
Add human formatting

ChrisPenner committed:
Fix migrating old reflog

ChrisPenner committed:
PR cleanup

ChrisPenner committed:
Refresh transcripts

ChrisPenner committed:
Merge remote-tracking branch 'origin/trunk' into cp/reflog-db

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Reflog.hs:
```diff
@@ -1,29 +0,0 @@
-{-# LANGUAGE ViewPatterns #-}
```
Everywhere just uses the `Reflog.Entry` type from `U.Codebase.Reflog` now.

ChrisPenner committed:
Docs and PR cleanup

aryairani reviewed (changes requested):
Awesome. A couple thoughts:

> Note that the new reflog table only has (timestamp, new root causal hash, reason); we no longer track a from causal, since it will always just be the causal before it, and we don't show this in the output anyways.

I think we still want to keep the old causal hash, just in case it somehow isn't the same as the entry before it. (The first entry has no entry before it, for example.)  Git is the same, in that it records both, but only shows the new hash in the output. We could also consider including both in the output (`#abc -> #def`), but it's messier from a "numbered output" perspective.

> We don't have timestamp info for old entries

Whoops!

> so we just import them all at the "current time", but add second offsets such that they appear in the correct chronological order.

Can there be a situation where I have 100 old entries and end up dating them into the future, and then a new user-generated entry is stuck in the middle of them (another case where I'd want my two causal hash ids per row again explicitly)? If it must be seconds and not nanoseconds or something (and even if it could be nanoseconds), putting them in the past would be safer.

> I don't delete the old reflog file, I could, but there doesn't seem to be a good reason to and deleting files on the user's system is always a little bit sketch.

:+1: If there's an easy spot to, you could tell the user we moved it and they can delete the file if they want.

ChrisPenner commented:
> I think we still want to keep the old causal hash, just in case it somehow isn't the same as the entry before it

I don't have any issue with adding the `from`, but how could this happen and what does it mean? (besides the first entry of course)
If it just means we have a bug and forgot to add something to the reflog I'd rather we fix that error rather than try to cover it up 😓 

If we want confidence in that behaviour we could try adding a "reason" to the actual "updateRoot" command rather than having a separate "appendReflog" command.

> Can there be a situation where I have 100 old entries and end up dating them into the future

No, the second offsets go into the past.

ChrisPenner committed:
Add 'from' to causals and append reflog within `putRootBranch`

ChrisPenner committed:
Add note about old reflog to reflog migration

ChrisPenner committed:
fixup! Add 'from' to causals and append reflog within `putRootBranch`

ChrisPenner commented:
@aryairani  

* I changed `putRootBranch` so it now accepts a reason and appends to the reflog in the same transaction which updates the root branch, so that should improve our confidence that the reflog is correct.
* I added the `from` causal to the reflog entries
* I added a note to the migration that the user can delete the old reflog file

~I just have to re-test things now 👍🏼~ ✅

aryairani commented:
> I don't have any issue with adding the `from`, but how could this happen and what does it mean? (besides the first entry of course). If it just means we have a bug and forgot to add something to the reflog I'd rather we fix that error rather than try to cover it up 😓

Yeah.  It shouldn't happen (except in that case), and we should fix whatever errors we find. This won't be covering them up, but it will help us not lose data in those cases. It's not the strongest rationale, I know.  I wonder what the reasons for git also including the old and new hashes in their reflog were too, could be interesting.

ChrisPenner committed:
Negate offset so things move into the past

aryairani commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs:
```diff
@@ -0,0 +1,64 @@
+      -- We offset existing entries by a number of seconds corresponding to their position in
+      -- the current file; we can't reclaim timestamps for old reflog entries, but this at
+      -- least puts them in the correct order chronologically.
+      let offsetTime = addUTCTime (negate $ fromInteger @NominalDiffTime n) now
```
as a sanity check now, wdyt about adding a screenshot of a migrated reflog demoing this?

aryairani replied:
~~I guess the transcript does enough here, so you can disregard.~~

aryairani replied:
(Actually disregard the disregard; the transcript is unrelated to reflog migration)

aryairani commented on unison-src/transcripts/reflog.output.md:
```diff
@@ -66,10 +66,10 @@ y = 2
-  2. #2a4hnm3mcj : add
-  3. #43lqh9lfp3 : builtins.merge
-  4. #sg60bvjo91 : (initial reflogged namespace)
+       When   Action           Root Hash
+  1.   now    add              #5fihla47kl
+  2.   now    add              #2a4hnm3mcj
+  3.   now    builtins.merge   #43lqh9lfp3
```
How hard would it be to restore the initial namespace to the output again?

aryairani commented:
Note: I could imagine having some reason to do multiple `putRootBranch` in a single command that should possibly just get a single reflog update, but this seems good for now.

aryairani reviewed (changes requested):
I left one request for display fix and one request for a sanity screenshot but otherwise looks good!

ChrisPenner committed:
Re-add initial and synthetic reflog entries

ChrisPenner commented:
I verified that the entries from my personal reflog are being displayed in the correct order.

Here's the beginning and end of history;

<img width="918" alt="image" src="https://user-images.githubusercontent.com/6439644/188018769-64a17f26-bc49-4614-a746-8b90ad3127f1.png">

<img width="924" alt="image" src="https://user-images.githubusercontent.com/6439644/188018797-24b32982-dda2-4a60-bbe5-401d2ff7e5f4.png">

ChrisPenner committed:
Reformat reflog output

aryairani commented:
Sweet!

aryairani commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs:
```diff
@@ -0,0 +1,64 @@
+    haveFrom <- isJust <$> Q.loadCausalByCausalHash (Reflog.fromRootCausalHash oldEntry)
+    haveTo <- isJust <$> Q.loadCausalByCausalHash (Reflog.toRootCausalHash oldEntry)
+    when (haveFrom && haveTo) $ Ops.appendReflog oldEntry
+  Sqlite.unsafeIO . putStrLn $ "I migrated old reflog entries from " <> reflogPath <> " into the codebase, you may delete that file now if you like."
```
```suggestion
  Sqlite.unsafeIO . putStrLn $ "I migrated old reflog entries from " <> reflogPath <> " into the codebase; you may delete that file now if you like."
```

aryairani committed:
Update parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs
